zebra-consensus/src/checkpoint.rs

```diff
@@ -3,315 +3,1062 @@
+        // The height of the last block we checked
+        let mut last_height = None;
+
+        for ((&height, qblocks), expected_height) in qrange.zip(expected_heights) {
+            // Check that the heights are continuous
+            if height != expected_height {
+                // There is a gap in the block chain
+                // Missing blocks are not an error - wait for more blocks
+                return (Some(expected_height), Some(expected_hash));
+            }
+
+            // Check if any queued block at this height is part of the chain
+            // TODO(teor): consider using find_map() with a qblock method
+            let parent_hash = qblocks
+                .iter()
+                .skip_while(|qblock| qblock.hash != expected_hash)
+                .map(|qblock| qblock.block.header.previous_block_hash)
+                .next();
+
+            if let Some(parent_hash) = parent_hash {
+                expected_hash = parent_hash;
+                last_height = Some(height);
+            } else {
+                // There was no valid block in the queued block list
+                // Missing blocks are not an error - wait for more blocks
+                //
+                // This check also handles zero-length vecs, which should be
+                // impossible
+                return (Some(expected_height), Some(expected_hash));
+            }
+        }
+
+        if let Some(last_height) = last_height {
+            // At this point, last_height is the height of the last block, but
+            // expected_hash is the hash of the last block's parent block.
+
+            // If we're at the genesis block, we want to return None for the parent
+            // block height.
+            let parent_height = last_height.0.checked_sub(1).map(BlockHeight);
+            (parent_height, Some(expected_hash))
+        } else {
+            // If the loop didn't execute, we're waiting for the next checkpoint block
+            // TODO(teor): add tests for this case
+            (Some(next_checkpoint_height), Some(next_checkpoint_hash))
+        }
+    }
+
+    /// Check if the current checkpoint range is complete.
+    ///
+    /// Verification begins with the genesis block, or the first block after
+    /// the previous checkpoint.
+    ///
+    /// Returns true if the range is complete, and false if more blocks
+    /// are needed to complete the range.
+    fn is_current_checkpoint_range_complete(&self) -> bool {
+        // The parent of the genesis block has height `None`.
+        let previous_checkpoint_height = self.previous_checkpoint_height();
+        // The genesis block's previous hash field is all zeroes
+        // TODO(teor): put this constant somewhere shared between both validators
+        let previous_checkpoint_hash = self
+            .previous_checkpoint_hash()
+            .unwrap_or(BlockHeaderHash([0; 32]));
+
+        // If checkpoint verification has finished, we should haved returned when
```

Comment:
```suggestion
        // If checkpoint verification has finished, we should have returned when
```

Reply:
Thanks!

I'm rewriting this code right now, I'll apply your fixes to these comments if they still exist after the rewrite :-)

Reply:
Both these comments were deleted or rewritten as part of the code rewrite. But thanks again!